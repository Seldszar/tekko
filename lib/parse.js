const { ParseError } = require("./errors");
const { unescapeString } = require("./helpers");

/**
 * Parses a message prefix.
 * @private
 * @param {String} input the message prefix input
 * @return {?MessagePrefix} the parsed message prefix.
 */
function parsePrefix(input) {
  if (!input) {
    return null;
  }

  let userIndex;
  let hostIndex;

  if ((userIndex = input.indexOf("!")) === 0) {
    return null;
  }

  if ((hostIndex = input.indexOf("@", userIndex + 1)) === 0) {
    return null;
  }

  const prefix = {
    name: null,
    user: null,
    host: null,
  };

  if (userIndex > -1 && hostIndex > -1) {
    prefix.name = input.slice(0, userIndex);
    prefix.user = input.slice(userIndex + 1, hostIndex);
    prefix.host = input.slice(hostIndex + 1);
  } else if (userIndex > -1) {
    prefix.name = input.slice(0, userIndex);
    prefix.user = input.slice(userIndex + 1);
  } else if (hostIndex > -1) {
    prefix.name = input.slice(0, hostIndex);
    prefix.host = input.slice(hostIndex + 1);
  } else {
    prefix.name = input;
  }

  return prefix;
}

/**
 * Parses a message.
 * @throws {ParseError}
 * @param {String} input the message input
 * @return {?Message} the parsed message.
 */
function parse(input) {
  if (!input) {
    throw new ParseError("Invalid message");
  }

  const message = {
    tags: {},
    prefix: null,
    command: null,
    params: [],
    trailing: null,
  };

  let position = 0;
  let nextWhitespace = 0;

  if (input.charCodeAt(0) === 64) {
    if ((nextWhitespace = input.indexOf(" ")) === -1) {
      throw new ParseError("Invalid message");
    }

    const tags = input.slice(1, nextWhitespace).split(";");
    const tagsLength = tags.length;

    for (let i = 0; i < tagsLength; i += 1) {
      const [key, value] = tags[i].split("=");
      message.tags[key] = value === undefined || unescapeString(value);
    }

    position = nextWhitespace + 1;
  }

  while (input.charCodeAt(position) === 32) {
    position += 1;
  }

  if (input.charCodeAt(position) === 58) {
    if ((nextWhitespace = input.indexOf(" ", position)) === -1) {
      throw new ParseError("Invalid message");
    }

    message.prefix = parsePrefix(input.slice(position + 1, nextWhitespace));
    position = nextWhitespace + 1;

    while (input.charCodeAt(position) === 32) {
      position += 1;
    }
  }

  if ((nextWhitespace = input.indexOf(" ", position)) === -1) {
    if (input.length > position) {
      message.command = input.slice(position);
      return message;
    }

    throw new ParseError("Invalid message");
  }

  message.command = input.slice(position, nextWhitespace);
  position = nextWhitespace + 1;

  while (input.charCodeAt(position) === 32) {
    position += 1;
  }

  while (position < input.length) {
    if (input.charCodeAt(position) === 58) {
      message.trailing = input.slice(position + 1);
      break;
    }

    if ((nextWhitespace = input.indexOf(" ", position)) === -1) {
      message.params.push(input.slice(position));
      break;
    }

    message.params.push(input.slice(position, nextWhitespace));
    position = nextWhitespace + 1;

    while (input.charCodeAt(position) === 32) {
      position += 1;
    }
  }

  return message;
}

module.exports = parse;
