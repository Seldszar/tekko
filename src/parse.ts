import { ParseError } from "./errors";
import { unescapeString } from "./helpers";
import { Message, MessagePrefix } from "./types";

/**
 * Parses a message prefix.
 * @param input the message prefix input
 * @return the parsed message prefix.
 */
function parsePrefix(input: string): MessagePrefix | null {
  if (!input) {
    return null;
  }

  let userIndex: number;
  let hostIndex: number;

  if ((userIndex = input.indexOf("!")) === 0) {
    return null;
  }

  if ((hostIndex = input.indexOf("@", userIndex + 1)) === 0) {
    return null;
  }

  const prefix: Partial<MessagePrefix> = {};

  if (userIndex > -1 && hostIndex > -1) {
    prefix.name = input.slice(0, userIndex);
    prefix.user = input.slice(userIndex + 1, hostIndex);
    prefix.host = input.slice(hostIndex + 1);
  } else if (userIndex > -1) {
    prefix.name = input.slice(0, userIndex);
    prefix.user = input.slice(userIndex + 1);
  } else if (hostIndex > -1) {
    prefix.name = input.slice(0, hostIndex);
    prefix.host = input.slice(hostIndex + 1);
  } else {
    prefix.name = input;
  }

  return prefix as MessagePrefix;
}

/**
 * Parses a message.
 * @throws {ParseError}
 * @param input the message input
 * @return the parsed message.
 */
export function parse(input: string): Message {
  const message: Partial<Message> = {
    params: [],
  };

  Object.defineProperties(message, {
    middle: {
      get(): string[] {
        return this.params.slice(0, -1);
      },
    },
    trailing: {
      get(): string {
        return this.params[this.params.length - 1];
      },
    },
  });

  let cursor = 0;
  let nextWhitespace = 0;

  if (input.charCodeAt(cursor) === 64) {
    if ((nextWhitespace = input.indexOf(" ")) === -1) {
      throw new ParseError("Invalid message");
    }

    message.tags = {};

    const tags = input.slice(1, nextWhitespace).split(";");
    const tagsLength = tags.length;

    for (let i = 0; i < tagsLength; i += 1) {
      const [key, value] = tags[i].split("=");
      message.tags[key] = value === undefined || unescapeString(value);
    }

    cursor = nextWhitespace + 1;
  }

  while (input.charCodeAt(cursor) === 32) {
    cursor += 1;
  }

  if (input.charCodeAt(cursor) === 58) {
    if ((nextWhitespace = input.indexOf(" ", cursor)) === -1) {
      throw new ParseError("Invalid message");
    }

    const prefix = parsePrefix(input.slice(cursor + 1, nextWhitespace));

    if (prefix) {
      message.prefix = prefix;
    }

    cursor = nextWhitespace + 1;

    while (input.charCodeAt(cursor) === 32) {
      cursor += 1;
    }
  }

  if ((nextWhitespace = input.indexOf(" ", cursor)) === -1) {
    if (input.length > cursor) {
      message.command = input.slice(cursor);
      return message as Message;
    }

    throw new ParseError("Invalid message");
  }

  message.command = input.slice(cursor, nextWhitespace);
  cursor = nextWhitespace + 1;

  while (input.charCodeAt(cursor) === 32) {
    cursor += 1;
  }

  while (cursor < input.length) {
    if (input.charCodeAt(cursor) === 58) {
      message.params.push(input.slice(cursor + 1));
      break;
    }

    if ((nextWhitespace = input.indexOf(" ", cursor)) === -1) {
      message.params.push(input.slice(cursor));
      break;
    }

    message.params.push(input.slice(cursor, nextWhitespace));
    cursor = nextWhitespace + 1;

    while (input.charCodeAt(cursor) === 32) {
      cursor += 1;
    }
  }

  return message as Message;
}
