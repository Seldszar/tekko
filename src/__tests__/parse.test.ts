import { parse as parseMessage } from "../parse";

describe("parse", () => {
  const testCases = [
    "",
    "@",
    "FOO",
    ":test ",
    ":test FOO",
    ":test FOO     ",
    ":! FOO",
    ":@ FOO",
    ":nick FOO",
    ":nick!user FOO",
    ":nick!user@host FOO",
    ":nick@host FOO",
    ":se.rv.er FOO",
    ":test!me@test.ing PRIVMSG #Test :This is a test",
    "PRIVMSG #Test :This is a test",
    ":test PRIVMSG foo :A string   with spaces   ",
    ":test     PRIVMSG    foo     :bar",
    ":test FOO bar baz quux",
    "FOO bar baz quux",
    "FOO    bar      baz   quux",
    "FOO bar baz quux :This is a test",
    ":test PRIVMSG #fo:oo :This is a test",
    "@test=super;single :test!me@test.ing FOO bar baz quux :This is a test",
    "@test=super;single     :test!me@test.ing    FOO   bar    baz   quux   :This is a test",
    "@test=super;single :  This is a test",
    "@test=super;single :",
    "@test=;single :",
    "@test=super\\shyper;single :",
    "@test=super\\r\\n\\\\\\:\\shyper\\;single :",
  ];

  for (const testCase of testCases) {
    test(`should parse "${testCase}"`, () => {
      try {
        const result = parseMessage(testCase);

        expect(result).toMatchSnapshot();
        expect(result.params).toMatchSnapshot();
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    });
  }

  test(`should parse with custom tag mapper`, () => {
    const result = parseMessage("@test=super;single :test!me@test.ing FOO :This is a test", {
      tagMapper(key, value) {
        return [`${key}_mapped`, `${value}_mapped`];
      },
    });

    expect(result).toMatchSnapshot();
  });
});
