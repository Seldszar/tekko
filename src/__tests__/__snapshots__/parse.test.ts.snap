// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse should parse "" 1`] = `[ParseError: Invalid message]`;

exports[`parse should parse ":! FOO" 1`] = `
Object {
  "command": "FOO",
  "params": Array [],
}
`;

exports[`parse should parse ":! FOO" 2`] = `Array []`;

exports[`parse should parse ":! FOO" 3`] = `undefined`;

exports[`parse should parse ":@ FOO" 1`] = `
Object {
  "command": "FOO",
  "params": Array [],
}
`;

exports[`parse should parse ":@ FOO" 2`] = `Array []`;

exports[`parse should parse ":@ FOO" 3`] = `undefined`;

exports[`parse should parse ":nick FOO" 1`] = `
Object {
  "command": "FOO",
  "params": Array [],
  "prefix": Object {
    "name": "nick",
  },
}
`;

exports[`parse should parse ":nick FOO" 2`] = `Array []`;

exports[`parse should parse ":nick FOO" 3`] = `undefined`;

exports[`parse should parse ":nick!user FOO" 1`] = `
Object {
  "command": "FOO",
  "params": Array [],
  "prefix": Object {
    "name": "nick",
    "user": "user",
  },
}
`;

exports[`parse should parse ":nick!user FOO" 2`] = `Array []`;

exports[`parse should parse ":nick!user FOO" 3`] = `undefined`;

exports[`parse should parse ":nick!user@host FOO" 1`] = `
Object {
  "command": "FOO",
  "params": Array [],
  "prefix": Object {
    "host": "host",
    "name": "nick",
    "user": "user",
  },
}
`;

exports[`parse should parse ":nick!user@host FOO" 2`] = `Array []`;

exports[`parse should parse ":nick!user@host FOO" 3`] = `undefined`;

exports[`parse should parse ":nick@host FOO" 1`] = `
Object {
  "command": "FOO",
  "params": Array [],
  "prefix": Object {
    "host": "host",
    "name": "nick",
  },
}
`;

exports[`parse should parse ":nick@host FOO" 2`] = `Array []`;

exports[`parse should parse ":nick@host FOO" 3`] = `undefined`;

exports[`parse should parse ":se.rv.er FOO" 1`] = `
Object {
  "command": "FOO",
  "params": Array [],
  "prefix": Object {
    "name": "se.rv.er",
  },
}
`;

exports[`parse should parse ":se.rv.er FOO" 2`] = `Array []`;

exports[`parse should parse ":se.rv.er FOO" 3`] = `undefined`;

exports[`parse should parse ":test     PRIVMSG    foo     :bar" 1`] = `
Object {
  "command": "PRIVMSG",
  "params": Array [
    "foo",
    "bar",
  ],
  "prefix": Object {
    "name": "test",
  },
}
`;

exports[`parse should parse ":test     PRIVMSG    foo     :bar" 2`] = `
Array [
  "foo",
]
`;

exports[`parse should parse ":test     PRIVMSG    foo     :bar" 3`] = `"bar"`;

exports[`parse should parse ":test " 1`] = `[ParseError: Invalid message]`;

exports[`parse should parse ":test FOO     " 1`] = `
Object {
  "command": "FOO",
  "params": Array [],
  "prefix": Object {
    "name": "test",
  },
}
`;

exports[`parse should parse ":test FOO     " 2`] = `Array []`;

exports[`parse should parse ":test FOO     " 3`] = `undefined`;

exports[`parse should parse ":test FOO bar baz quux" 1`] = `
Object {
  "command": "FOO",
  "params": Array [
    "bar",
    "baz",
    "quux",
  ],
  "prefix": Object {
    "name": "test",
  },
}
`;

exports[`parse should parse ":test FOO bar baz quux" 2`] = `
Array [
  "bar",
  "baz",
]
`;

exports[`parse should parse ":test FOO bar baz quux" 3`] = `"quux"`;

exports[`parse should parse ":test FOO" 1`] = `
Object {
  "command": "FOO",
  "params": Array [],
  "prefix": Object {
    "name": "test",
  },
}
`;

exports[`parse should parse ":test FOO" 2`] = `Array []`;

exports[`parse should parse ":test FOO" 3`] = `undefined`;

exports[`parse should parse ":test PRIVMSG #fo:oo :This is a test" 1`] = `
Object {
  "command": "PRIVMSG",
  "params": Array [
    "#fo:oo",
    "This is a test",
  ],
  "prefix": Object {
    "name": "test",
  },
}
`;

exports[`parse should parse ":test PRIVMSG #fo:oo :This is a test" 2`] = `
Array [
  "#fo:oo",
]
`;

exports[`parse should parse ":test PRIVMSG #fo:oo :This is a test" 3`] = `"This is a test"`;

exports[`parse should parse ":test PRIVMSG foo :A string   with spaces   " 1`] = `
Object {
  "command": "PRIVMSG",
  "params": Array [
    "foo",
    "A string   with spaces   ",
  ],
  "prefix": Object {
    "name": "test",
  },
}
`;

exports[`parse should parse ":test PRIVMSG foo :A string   with spaces   " 2`] = `
Array [
  "foo",
]
`;

exports[`parse should parse ":test PRIVMSG foo :A string   with spaces   " 3`] = `"A string   with spaces   "`;

exports[`parse should parse ":test!me@test.ing PRIVMSG #Test :This is a test" 1`] = `
Object {
  "command": "PRIVMSG",
  "params": Array [
    "#Test",
    "This is a test",
  ],
  "prefix": Object {
    "host": "test.ing",
    "name": "test",
    "user": "me",
  },
}
`;

exports[`parse should parse ":test!me@test.ing PRIVMSG #Test :This is a test" 2`] = `
Array [
  "#Test",
]
`;

exports[`parse should parse ":test!me@test.ing PRIVMSG #Test :This is a test" 3`] = `"This is a test"`;

exports[`parse should parse "@" 1`] = `[ParseError: Invalid message]`;

exports[`parse should parse "@test=;single :" 1`] = `[ParseError: Invalid message]`;

exports[`parse should parse "@test=super;single     :test!me@test.ing    FOO   bar    baz   quux   :This is a test" 1`] = `
Object {
  "command": "FOO",
  "params": Array [
    "bar",
    "baz",
    "quux",
    "This is a test",
  ],
  "prefix": Object {
    "host": "test.ing",
    "name": "test",
    "user": "me",
  },
  "tags": Object {
    "single": true,
    "test": "super",
  },
}
`;

exports[`parse should parse "@test=super;single     :test!me@test.ing    FOO   bar    baz   quux   :This is a test" 2`] = `
Array [
  "bar",
  "baz",
  "quux",
]
`;

exports[`parse should parse "@test=super;single     :test!me@test.ing    FOO   bar    baz   quux   :This is a test" 3`] = `"This is a test"`;

exports[`parse should parse "@test=super;single :  This is a test" 1`] = `
Object {
  "command": "This",
  "params": Array [
    "is",
    "a",
    "test",
  ],
  "tags": Object {
    "single": true,
    "test": "super",
  },
}
`;

exports[`parse should parse "@test=super;single :  This is a test" 2`] = `
Array [
  "is",
  "a",
]
`;

exports[`parse should parse "@test=super;single :  This is a test" 3`] = `"test"`;

exports[`parse should parse "@test=super;single :" 1`] = `[ParseError: Invalid message]`;

exports[`parse should parse "@test=super;single :test!me@test.ing FOO bar baz quux :This is a test" 1`] = `
Object {
  "command": "FOO",
  "params": Array [
    "bar",
    "baz",
    "quux",
    "This is a test",
  ],
  "prefix": Object {
    "host": "test.ing",
    "name": "test",
    "user": "me",
  },
  "tags": Object {
    "single": true,
    "test": "super",
  },
}
`;

exports[`parse should parse "@test=super;single :test!me@test.ing FOO bar baz quux :This is a test" 2`] = `
Array [
  "bar",
  "baz",
  "quux",
]
`;

exports[`parse should parse "@test=super;single :test!me@test.ing FOO bar baz quux :This is a test" 3`] = `"This is a test"`;

exports[`parse should parse "@test=super\\r\\n\\\\\\:\\shyper\\;single :" 1`] = `[ParseError: Invalid message]`;

exports[`parse should parse "@test=super\\shyper;single :" 1`] = `[ParseError: Invalid message]`;

exports[`parse should parse "FOO    bar      baz   quux" 1`] = `
Object {
  "command": "FOO",
  "params": Array [
    "bar",
    "baz",
    "quux",
  ],
}
`;

exports[`parse should parse "FOO    bar      baz   quux" 2`] = `
Array [
  "bar",
  "baz",
]
`;

exports[`parse should parse "FOO    bar      baz   quux" 3`] = `"quux"`;

exports[`parse should parse "FOO bar baz quux :This is a test" 1`] = `
Object {
  "command": "FOO",
  "params": Array [
    "bar",
    "baz",
    "quux",
    "This is a test",
  ],
}
`;

exports[`parse should parse "FOO bar baz quux :This is a test" 2`] = `
Array [
  "bar",
  "baz",
  "quux",
]
`;

exports[`parse should parse "FOO bar baz quux :This is a test" 3`] = `"This is a test"`;

exports[`parse should parse "FOO bar baz quux" 1`] = `
Object {
  "command": "FOO",
  "params": Array [
    "bar",
    "baz",
    "quux",
  ],
}
`;

exports[`parse should parse "FOO bar baz quux" 2`] = `
Array [
  "bar",
  "baz",
]
`;

exports[`parse should parse "FOO bar baz quux" 3`] = `"quux"`;

exports[`parse should parse "FOO" 1`] = `
Object {
  "command": "FOO",
  "params": Array [],
}
`;

exports[`parse should parse "FOO" 2`] = `Array []`;

exports[`parse should parse "FOO" 3`] = `undefined`;

exports[`parse should parse "PRIVMSG #Test :This is a test" 1`] = `
Object {
  "command": "PRIVMSG",
  "params": Array [
    "#Test",
    "This is a test",
  ],
}
`;

exports[`parse should parse "PRIVMSG #Test :This is a test" 2`] = `
Array [
  "#Test",
]
`;

exports[`parse should parse "PRIVMSG #Test :This is a test" 3`] = `"This is a test"`;
